go-controller/pkg/ovn/external_ids_syncer/port_group/port_group_sync.go

```diff
@@ -0,0 +1,230 @@
+	// Set to 0 to disable batching
+	txnBatchSize int
+}
+
+func getControllerName(networkExternalID string) string {
+	if networkExternalID == "" {
+		return defaultNetworkControllerName
+	}
+	return networkExternalID + "-network-controller"
+}
+
+// NewPortGroupSyncer creates a PortGroupSyncer that will sync port groups without the new ExternalIDs for all
+// controllers. Controller name will be defined based on getControllerName() function.
+func NewPortGroupSyncer(nbClient libovsdbclient.Client) *PortGroupSyncer {
+	return &PortGroupSyncer{
+		nbClient:     nbClient,
+		txnBatchSize: 50,
+	}
+}
+
+func getPortGroupNamespaceDbIDs(namespace, networkExternalID string) *libovsdbops.DbObjectIDs {
+	controllerName := getControllerName(networkExternalID)
+	return libovsdbops.NewDbObjectIDs(libovsdbops.PortGroupNamespace, controllerName, map[libovsdbops.ExternalIDKey]string{
+		libovsdbops.ObjectNameKey: namespace,
+	})
+}
+
+func getPortGroupNetpolNamespaceDbIDs(namespace, direction, networkExternalID string) *libovsdbops.DbObjectIDs {
+	controllerName := getControllerName(networkExternalID)
+	return libovsdbops.NewDbObjectIDs(libovsdbops.PortGroupNetpolNamespace, controllerName, map[libovsdbops.ExternalIDKey]string{
+		libovsdbops.ObjectNameKey:      namespace,
+		libovsdbops.PolicyDirectionKey: direction,
+	})
+}
+
+func getPortGroupNetworkPolicyDbIDs(policyNamespace, policyName, networkExternalID string) *libovsdbops.DbObjectIDs {
+	controllerName := getControllerName(networkExternalID)
+	return libovsdbops.NewDbObjectIDs(libovsdbops.PortGroupNetworkPolicy, controllerName,
+		map[libovsdbops.ExternalIDKey]string{
+			libovsdbops.ObjectNameKey: fmt.Sprintf("%s_%s", policyNamespace, policyName),
+		})
+}
+
+func getPortGroupClusterDbIDs(baseName, networkExternalID string) *libovsdbops.DbObjectIDs {
+	controllerName := getControllerName(networkExternalID)
+	return libovsdbops.NewDbObjectIDs(libovsdbops.PortGroupCluster, controllerName, map[libovsdbops.ExternalIDKey]string{
+		libovsdbops.ObjectNameKey: baseName,
+	})
+}
+
+// getReferencingObjsAndNewDbIDs finds all object that reference stale port group and tries to create a new dbIDs
+// based on referencing objects
+func (syncer *PortGroupSyncer) getReferencingObjsAndNewDbIDs(oldHash, oldName, networkExternalID string) (acls []*nbdb.ACL,
+	dbIDs *libovsdbops.DbObjectIDs, err error) {
+	// find all referencing objects
+	aclPred := func(acl *nbdb.ACL) bool {
+		return strings.Contains(acl.Match, "@"+oldHash)
+	}
+	acls, err = libovsdbops.FindACLsWithPredicate(syncer.nbClient, aclPred)
+	if err != nil {
+		err = fmt.Errorf("failed to find acls for port group %s: %v", oldHash, err)
+		return
+	}
+	// build dbIDs
+	switch {
+	// Filter port groups with pre-defined names
+	case oldName == types.ClusterPortGroupNameBase || oldName == types.ClusterRtrPortGroupNameBase:
+		dbIDs = getPortGroupClusterDbIDs(oldName, networkExternalID)
+	case strings.Contains(oldName, "_"):
```

Comment:
can't the underscore just be part of a namespace name?

Reply:
nope, only dashes are allowed in namespace names https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#dns-label-names